# Проект по А/B-тестированию

Ваша задача — провести оценку результатов A/B-теста. В вашем распоряжении
есть датасет с действиями пользователей, техническое задание и несколько
вспомогательных датасетов.
Оцените корректность проведения теста и проанализируйте его результаты.
Чтобы оценить корректность проведения теста:
удостоверьтесь, что нет пересечений с конкурирующим тестом и нет
пользователей, участвующих в двух группах теста одновременно;
проверьте равномерность распределения пользователей по тестовым группам
и правильность их формирования.

## Описание данных

- ab_project_marketing_events.csv — календарь маркетинговых событий на 2020 год.
    - Структура файла:
        - name — название маркетингового события;
        - regions — регионы, в которых будет проводиться рекламная кампания;
        - start_dt — дата начала кампании;
        - finish_dt — дата завершения кампании.
        <br>
        <br>
- final_ab_new_users.csv — пользователи, зарегистрировавшиеся с 7 по 21 декабря 2020 года. 
    - Структура файла:
        - user_id — идентификатор пользователя;
        - first_date — дата регистрации;
        - region — регион пользователя;
        - device — устройство, с которого происходила регистрация.
        <br>
        <br>
- final_ab_events.csv — действия новых пользователей в период с 7 декабря 2020 по 4 января 2021 года. 
    - Структура файла:
        - user_id — идентификатор пользователя;
        - event_dt — дата и время покупки;
        - event_name — тип события;
        - details — дополнительные данные о событии. Например, для покупок, purchase, в этом поле хранится стоимость покупки в долларах.
        <br>
        <br>
- final_ab_participants.csv — таблица участников тестов. 
    - Структура файла:
        - user_id — идентификатор пользователя;
        - ab_test — название теста;
        - group — группа пользователя.

## План проекта

- Опишите цели исследования
- Исследуйте данные:
    - Требуется ли преобразование типов?
    - Опишите природу пропущенных значений и дубликатов, если их обнаружите.
- Оцените корректность проведения теста. Обратите внимание на:
    - Соответствие данных требованиям технического задания. Проверьте корректность всех пунктов технического задания.
    - Время проведения теста. Убедитесь, что оно не совпадает с маркетинговыми и другими активностями.
    - Аудиторию теста. Удостоверьтесь, что нет пересечений с конкурирующим тестом и нет пользователей, участвующих в двух группах - теста одновременно. Проверьте равномерность распределения по тестовым группам и правильность их формирования.
- Проведите исследовательский анализ данных:
    - Количество событий на пользователя одинаково распределены в выборках?
    - Как число событий в выборках распределено по дням?
    - Как меняется конверсия в воронке в выборках на разных этапах?
    - Какие особенности данных нужно учесть, прежде чем приступать к A/B-тестированию?
- Оцените результаты A/B-тестирования:
    - Что можно сказать про результаты A/В-тестирования?
    - Проверьте статистическую разницу долей z-критерием.
- Опишите выводы по этапу исследовательского анализа данных и по проведённой оценке результатов A/B-тестирования. Сделайте общее заключение о корректности проведения теста.

## Знакомство с данными

In [1]:
import pandas as pd

In [2]:
marketing_events = pd.read_csv('https://code.s3.yandex.net/datasets/ab_project_marketing_events.csv')
new_users = pd.read_csv('https://code.s3.yandex.net/datasets/final_ab_new_users.csv')
events = pd.read_csv('https://code.s3.yandex.net/datasets/final_ab_events.csv')
participants = pd.read_csv('https://code.s3.yandex.net/datasets/final_ab_participants.csv')

In [3]:
marketing_events

name                   regions    start_dt  \
0           Christmas&New Year Promo             EU, N.America  2020-12-25   
1       St. Valentine's Day Giveaway  EU, CIS, APAC, N.America  2020-02-14   
2             St. Patric's Day Promo             EU, N.America  2020-03-17   
3                       Easter Promo  EU, CIS, APAC, N.America  2020-04-12   
4                  4th of July Promo                 N.America  2020-07-04   
5          Black Friday Ads Campaign  EU, CIS, APAC, N.America  2020-11-26   
6             Chinese New Year Promo                      APAC  2020-01-25   
7   Labor day (May 1st) Ads Campaign             EU, CIS, APAC  2020-05-01   
8    International Women's Day Promo             EU, CIS, APAC  2020-03-08   
9    Victory Day CIS (May 9th) Event                       CIS  2020-05-09   
10         CIS New Year Gift Lottery                       CIS  2020-12-30   
11     Dragon Boat Festival Giveaway                      APAC  2020-06-25   
12           Single's Day Gift Promo                      APAC  2020-11-11   
13             Chinese Moon Festival                      APAC  2020-10-01   

     finish_dt  
0   2021-01-03  
1   2020-02-16  
2   2020-03-19  
3   2020-04-19  
4   2020-07-11  
5   2020-12-01  
6   2020-02-07  
7   2020-05-03  
8   2020-03-10  
9   2020-05-11  
10  2021-01-07  
11  2020-07-01  
12  2020-11-12  
13  2020-10-07

In [4]:
marketing_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   name       14 non-null     object
 1   regions    14 non-null     object
 2   start_dt   14 non-null     object
 3   finish_dt  14 non-null     object
dtypes: object(4)
memory usage: 580.0+ bytes


In [5]:
marketing_events.isna().sum()


name         0
regions      0
start_dt     0
finish_dt    0
dtype: int64

In [6]:
marketing_events.duplicated().sum()

0

In [7]:
new_users

user_id  first_date     region   device
0      D72A72121175D8BE  2020-12-07         EU       PC
1      F1C668619DFE6E65  2020-12-07  N.America  Android
2      2E1BF1D4C37EA01F  2020-12-07         EU       PC
3      50734A22C0C63768  2020-12-07         EU   iPhone
4      E1BDDCE0DAFA2679  2020-12-07  N.America   iPhone
...                 ...         ...        ...      ...
61728  1DB53B933257165D  2020-12-20         EU  Android
61729  538643EB4527ED03  2020-12-20         EU      Mac
61730  7ADEE837D5D8CBBD  2020-12-20         EU       PC
61731  1C7D23927835213F  2020-12-20         EU   iPhone
61732  8F04273BB2860229  2020-12-20         EU  Android

[61733 rows x 4 columns]

In [8]:
new_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61733 entries, 0 to 61732
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     61733 non-null  object
 1   first_date  61733 non-null  object
 2   region      61733 non-null  object
 3   device      61733 non-null  object
dtypes: object(4)
memory usage: 1.9+ MB


In [9]:
new_users.isna().sum()

user_id       0
first_date    0
region        0
device        0
dtype: int64

In [10]:
new_users.duplicated().sum()

0

In [11]:
events

user_id             event_dt event_name  details
0       E1BDDCE0DAFA2679  2020-12-07 20:22:03   purchase    99.99
1       7B6452F081F49504  2020-12-07 09:22:53   purchase     9.99
2       9CD9F34546DF254C  2020-12-07 12:59:29   purchase     4.99
3       96F27A054B191457  2020-12-07 04:02:40   purchase     4.99
4       1FD7660FDF94CA1F  2020-12-07 10:15:09   purchase     4.99
...                  ...                  ...        ...      ...
440312  245E85F65C358E08  2020-12-30 19:35:55      login      NaN
440313  9385A108F5A0A7A7  2020-12-30 10:54:15      login      NaN
440314  DB650B7559AC6EAC  2020-12-30 10:59:09      login      NaN
440315  F80C9BDDEA02E53C  2020-12-30 09:53:39      login      NaN
440316  7AEC61159B672CC5  2020-12-30 11:36:13      login      NaN

[440317 rows x 4 columns]

In [12]:
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440317 entries, 0 to 440316
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   user_id     440317 non-null  object 
 1   event_dt    440317 non-null  object 
 2   event_name  440317 non-null  object 
 3   details     62740 non-null   float64
dtypes: float64(1), object(3)
memory usage: 13.4+ MB


In [13]:
events.isna().sum()

user_id            0
event_dt           0
event_name         0
details       377577
dtype: int64

In [14]:
events.duplicated().sum()

0

In [15]:
participants

user_id group                  ab_test
0      D1ABA3E2887B6A73     A  recommender_system_test
1      A7A3664BD6242119     A  recommender_system_test
2      DABC14FDDFADD29E     A  recommender_system_test
3      04988C5DF189632E     A  recommender_system_test
4      482F14783456D21B     B  recommender_system_test
...                 ...   ...                      ...
18263  1D302F8688B91781     B        interface_eu_test
18264  3DE51B726983B657     A        interface_eu_test
18265  F501F79D332BE86C     A        interface_eu_test
18266  63FBE257B05F2245     A        interface_eu_test
18267  79F9ABFB029CF724     B        interface_eu_test

[18268 rows x 3 columns]

In [16]:
participants.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18268 entries, 0 to 18267
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  18268 non-null  object
 1   group    18268 non-null  object
 2   ab_test  18268 non-null  object
dtypes: object(3)
memory usage: 428.3+ KB


In [17]:
participants.isna().sum()

user_id    0
group      0
ab_test    0
dtype: int64

In [18]:
participants.duplicated().sum()

0

## Иследование данных

### Преобразование типов

В таблицах marketing_events, events, new_users все столбцы с датами имеют не правильный тип данных. Меняем. 

In [19]:
marketing_events['start_dt'] = pd.to_datetime(marketing_events['start_dt'])
marketing_events['finish_dt'] = pd.to_datetime(marketing_events['finish_dt'])
new_users['first_date'] = pd.to_datetime(new_users['first_date'])
events['event_dt'] = pd.to_datetime(events['event_dt'])

Пропусков и дубликатов ни одна из таблиц не имееет. Переходим к следующему шагу иследования.

## Оценка корректности анализа

### Требования по ТЗ

1. 15% новых пользователей должны быть из региона EU
2. Зарегистрированные с 07.12.20 по 21.12.20
3. Продолжительность теста 07.12.20 по 04.01.21
4. Тест recomented_system_test

### Регион